File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self, root: Optional[TreeNode], min: int = -inf, max: int = inf) -> bool:
        if not root:
            return True
        if root.val > min and root.val < max:
            return self.dfs(root.left, min, root.val) and self.dfs(root.right, root.val, max)
        else:
            return False
        
    def bfs(self, root: Optional[TreeNode]) -> bool:
        q = deque([[-inf, root, inf]])
        while q:
            mn, root, mx = q.popleft()
            if root.val <= mn or root.val >= mx:
                return False
            if root.left:
                q.append([mn, root.left, root.val])
            if root.right:
                q.append([root.val, root.right, mx])
        return True
    
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.bfs(root)
        
        #
        
        return self.dfs(root)

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isValidBST(root: TreeNode | null): boolean {
    if (!root) return true
    const stack: TreeNode[] = []
    let parent: TreeNode | null = null
    while (stack.length || root) {
        if (root) {
            stack.push(root)
            root = root.left
        } else {
            root = stack.pop()
            if (parent && root.val <= parent.val) return false
            parent = root
            root = root.right
        }
    }
    return true
    
    //
    
    return dfs(root, -Infinity, Infinity)
    
    function dfs(root: TreeNode | null, lo: number, hi: number): boolean {
        if (!root) return true
        if (root.val <= lo || root.val >= hi) return false
        return dfs(root.left, lo, root.val) && dfs(root.right, root.val, hi)
    }
};